devicetree: add dependency information to edtlib

pabigot opened the pull request:
Device tree nodes have a dependency on their parents, as well as other nodes.  To ensure driver instances are initialized in the proper we need to identify these dependencies and construct a total order on nodes that ensures no node is initialized before something it depends on.

Add a Graph class that calculates a partial order on nodes, taking into account the potential for cycles in the dependency graph.

When generating devicetree value headers calculate the dependency graph and document the order and dependencies in the derived files. In the future these dependencies may be expressed in binding data.

Example of content added by this:
```
# Devicetree node: /soc/gpio@50000000
# Binding (compatible = nordic,nrf-gpio): /mnt/devel/external/zp/zephyr/dts/bindings/gpio/nordic,nrf-gpio.yaml
# Binding description: This is a representation of the NRF GPIO nodes
# Ordinal: 4
# Requires: 3 /soc
# Supports: 5 /buttons/button_0
# Supports: 27 /soc/i2c@40003000/hts221@5f
...
# Devicetree node: /soc/i2c@40003000/hts221@5f
# Binding (compatible = st,hts221): /mnt/devel/external/zp/zephyr/dts/bindings/sensor/st,hts221.yaml
# Binding description: This binding gives a base representation of HTS221 humidity and temperature sensor
# Ordinal: 27
# Requires: 4 /soc/gpio@50000000
# Requires: 10 /soc/i2c@40003000
```

pabigot committed:
scripts: dts: add dependency information to edtlib

Device tree nodes have a dependency on their parents, as well as other
nodes.  To ensure driver instances are initialized in the proper we
need to identify these dependencies and construct a total order on
nodes that ensures no node is initialized before something it depends
on.

Add a Graph class that calculates a partial order on nodes, taking
into account the potential for cycles in the dependency graph.

When generating devicetree value headers calculate the dependency
graph and document the order and dependencies in the derived files.
In the future these dependencies may be expressed in binding data.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pizi-nordic reviewed (approved):
I fully support all work towards proper handling of dependencies in Zephyr :)

pizi-nordic commented:
FYI: @barsok

ulfalizer commented on scripts/dts/testedtlib.py:
```diff
@@ -199,8 +199,25 @@ def verify_streq(actual, expected):
+    verify_eq(edt.get_node("/in-dir-1").ordinal, 1)
+    if edt.get_node("/") not in edt.depends_on(edt.get_node("/in-dir-1")):
+        fail("/ depends-on /in-dir-1")
+    if edt.get_node("/in-dir-1")not in edt.required_by(edt.get_node("/")):
```
Missing space before `not`.

ulfalizer reviewed (changes requested):
Some nits and a `defaultdict` tip.

Feels like a simpler algo might be possible if all that's needed is ordering + dependency loop detection, but shouldn't hurt at least.

pabigot commented:
@ulfalizer OK, done with updates, please re-check.

Many of your comments (style, features) are due to "grutil" being derived from a more general dependency module I wrote in Python 2 over ten years ago.  I've done what I can to convert the code to your preferred style.

One thing I have not done is changed the term `ordinal`.  At the moment, this PR introduces the only order on EDT Nodes present in the system: a stable total order that satisfies the relations in the dependency graph.

I don't want to call this "init(ialization) order" because that's not what it is.  Zephyr partitions initialization into four phases: `PRE_KERNEL_`, `PRE_KERNEL_2`, `POST_KERNEL`, and `APPLICATION`.  Initialization order is currently defined by Kconfig priorities within each group.

There is a relation between initialization order and the partial order defined by dependencies, and that order should ultimately be provided by the devicetree scripting, but it won't be the same as this order.

I also don't want to call the Node attribute "order" because that's not what it is either.  It's the position of a node within a sequence, i.e. its ordinal.

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -141,6 +144,58 @@ def chosen_node(self, name):
+
+        # Calculate an order that ensures no node is before any node
+        # it depends on.
+        self._graph.scc_order()
```
Still need to check for the exception from `_tarjan()` here, and convert it to an `EDTError`.

ulfalizer commented:
> One thing I have not done is changed the term `ordinal`. At the moment, this PR introduces the only order on EDT Nodes present in the system: a stable total order that satisfies the relations in the dependency graph.

What about something like `dep_ordinal`, to give some hint at least? Even if you know what "ordinal" means (I was shaky on it), it's not obvious what the sequence is.

Personally, I try hard to avoid using a word/term when I suspect most people might not know what it means. It's a bit more effort and less auto-pilot, but can make stuff more efficient for people reading the code.

pabigot commented:
@ulfalizer ok exception handling and attribute rename done.

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -141,6 +144,62 @@ def chosen_node(self, name):
+
+        # Calculate an order that ensures no node is before any node
+        # it depends on.
+        return self.scc_order()
```
Could do something like this to convert it:

```python
try:
    self.scc_order()
except Exception as e:
    raise EDTError(str(e))
```

Maybe a more specific exception could be raised from `_tarjan()` too, but won't nit that hard.

pabigot replied:
Note that the exception translation is done in the `scc_order` method, which I've updated to look like this.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -58,6 +67,14 @@ def main():
                         blank_before=False)
             out_comment("Binding description: " + node.description,
                         blank_before=False)
+            out_comment("Ordinal: {}".format(node.dep_ordinal),
```
Could say `Dependency ordinal: ` maybe.

ulfalizer reviewed (approved):
Markup in docstrings and `__` variables aren't used elsewhere, so that's a bit inconsistent, but not super important.

ulfalizer commented on scripts/dts/testedtlib.py:
```diff
@@ -199,8 +199,25 @@ def verify_streq(actual, expected):
+    if edt.get_node("/in-dir-1") not in edt.required_by(edt.get_node("/")):
+        fail("/in-dir-1 required-by /")
+
+    print("all tests passed")
```
Could add one more blank line here for some PEP 8 zealotry (two blanks between top-level functions/classes).

ulfalizer commented on scripts/dts/grutils.py:
```diff
@@ -0,0 +1,150 @@
+                    self.__roots.add(n)
+        return self.__roots
+
+    def nodes (self):
```
Could have a `@property` here, so that the syntax becomes `graph.nodes` instead of  `graph.nodes()`.

ulfalizer commented on scripts/dts/grutils.py:
```diff
@@ -0,0 +1,143 @@
+        # dependency cycle, and to impose a total order on components
+        # based on dependencies.
+
+        self.__scc_map = {}
```
This variable seems to be unused.

ulfalizer commented:
Remembered another thing.

Might be good to generate edges for `Node.interrupts` as well. That one's still hardcoded, because it doesn't work like any other property (and deals with both `interrupts` and `interrupts-extended`).

pabigot commented:
Rebased and updated generated text to be consistent with new documentation format:
```
/*
 * Devicetree node:
 *   /soc/gpio@50000000
 *
 * Binding (compatible = nordic,nrf-gpio):
 *   $ZEPHYR_BASE/dts/bindings/gpio/nordic,nrf-gpio.yaml
 *
 * Dependency Ordinal: 4
 *
 * Requires:
 *   3 /soc
 *
 * Supports:
 *   5 /buttons/button_0
 *   28 /soc/i2c@40003000/hts221@5f
 *   31 /soc/i2c@40004000/lis2dh12@19
 *
 * Description:
 *   This is a representation of the NRF GPIO nodes
 */
```
